File: dashboard/lib/replay.py
"""Night replay (Task 11 / design record §4) — a journal time-window becomes ordered field frames.

The replay is a beloved **treat, never load-bearing** (§0.1 owner ruling, §4): it plays back any
window of the append-only ``journal.jsonl`` as a scrubbable, steppable time-lapse over the SAME
airfield engine the live field uses, every frame clickable through to the raw journal record that
produced it. It answers "watch what the field did" — never "what do I need to do now" (that is the
mechanical digest's job, and the live Needs You panel's).

Everything here is a PURE function of a journal (a list of records) plus an injected ``hhmm``
formatter — no wall clock, no I/O, no ``gh`` — so the whole reconstruction is unit-tested to the
line. The load-bearing disciplines:

* **Frames run forward in time (§4).** The window is sorted by ``ts``; a record with no usable ts
  (a corrupt ``NaN``, a half-written line) can't be placed on a timeline and is dropped — the movie
  never jumps.
* **Position is DISCRETE and derived through the truth layer (§3, B.1).** Each flight's stage at a
  frame comes from :func:`flights.flight_stage` fed by a tiny per-flight fact accumulator walked
  from journal landmarks — the SAME function the live field derives its stage from. The replay is
  the state diagram in motion, not a second model that could drift (the squint test). A plane only
  moves when a real journal event changed its stage, and that transit IS the event you are watching.
* **The living clock is honest per frame (§7).** Each frame's lighting (``daypart``) is the wall
  clock at THAT event's real ts — a 2 a.m. park replays under night lighting — never "now".
* **Every frame is a click target (§4).** It carries the offending flight ``num``, the glossed
  comms sentence (via the tested :mod:`tower` vocabulary), and the exact ``raw`` journal line, so a
  tap on any frame opens that flight's drawer or expands ground truth.
"""
import json
import math
import time

import flights
import tower


# Air working stages whose planes trail a contrail in the replay (a treat has no liveness axis, so
# the contrail is a plain "is this plane flying a working leg" flag — the engine only draws trails
# for air anchors anyway). A grey session-frozen / amber awaiting hull never trails (§5).
_TRAILING = {flights.TAKEOFF, flights.DOWNWIND, flights.BASE_TURN, flights.FINAL, flights.HOLDING}

# The most frames a single replay ships by default — a localhost treat, but an unbounded journal
# should never balloon the payload. When the window is bigger, the MOST RECENT frames are kept (the
# recent past is what you replay) and the window is flagged truncated so the client can say so.
_DEFAULT_MAX_FRAMES = 600


def _finite(v):
    return isinstance(v, (int, float)) and not isinstance(v, bool) and math.isfinite(v)


def _local_hhmm(ts):
    """Local wall-clock ``HH:MM`` for a ts, or ``""`` when unusable — the same guarded formatting
    as the server's, so a lib-only caller still gets sensible times. Injectable (the server passes
    its own; tests pass a deterministic stub) so timezone never leaks into an assertion."""
    if not _finite(ts):
        return ""
    try:
        return time.strftime("%H:%M", time.localtime(ts))
    except (ValueError, OSError, OverflowError):
        return ""


def _daypart(ts):
    """The frame's lighting for a ts — :func:`flights.daypart`, but guarded. A corrupt-but-FINITE
    ts (a huge or negative epoch that slips past the ``NaN``/``Infinity`` screen) overflows
    ``time.localtime``; the replay must never 500 on one bad journal line (the derivations-never-raise
    rule), so unusable ts fall back to plain ``day`` lighting."""
    try:
        return flights.daypart(ts)
    except (ValueError, OSError, OverflowError):
        return "day"


# =============================== per-flight fact accumulator ===============================
# The replay reconstructs each flight's DISCRETE facts by walking its journal landmarks in time
# order, then hands them to the tested flights.flight_stage — never inventing a parallel stage map.

def _new_facts():
    """A flight's blank slate before its first event."""
    return {"status": None, "launched": False, "session_started": False,
            "report": False, "cleared": False, "closed": False, "attempt": 1}


def _apply(facts, rec):
    """Fold one journal record into a flight's facts. Only records that move the DISCRETE position
    change anything; flavor acts (nudge, notify, relabel, answers) leave the stage exactly as it
    was — the plane holds its anchor, honest to the journal."""
    act = rec.get("act")
    if act == "launch":
        facts["launched"] = True
        facts["session_started"] = False       # just off the stand — a brief takeoff before the leg
        facts["status"] = "running"
    elif act == "regenerate":
        # A go-around is an honest retire-and-rebuild (§3): a NEW attempt taxis out from the start.
        facts["attempt"] += 1
        facts["launched"] = True
        facts["session_started"] = False
        facts["report"] = False
        facts["cleared"] = False
        facts["status"] = "running"
    elif act == "gate":
        if rec.get("outcome") == "ok":
            facts["cleared"] = True
            facts["status"] = "gating"
        # a held/failed gate is NOT a pass — the plane stays at its report/base position
    elif act == "merge":
        if rec.get("outcome") == "ok":
            facts["status"] = "merged"          # the landing; a missed approach (failed merge) holds
    elif act == "park":
        # park carries the whole "machine gave up" family; needs_william is the amber owner-decision.
        facts["status"] = "needs_william" if rec.get("needs_william") else "parked"
    elif act == "bounce":
        facts["status"] = "bounced"
    elif act == "hold":
        facts["status"] = "holding"
    elif act == "drop":
        facts["closed"] = True
    elif act == "event":
        ev = rec.get("event") if isinstance(rec.get("event"), dict) else {}
        et = ev.get("type")
        if et == "session_finished":
            facts["report"] = True
            facts["session_started"] = True
            facts["status"] = "running"
        elif et == "frozen":
            facts["status"] = "frozen"
        elif et == "exited":
            facts["status"] = "exited"
        elif et == "session_blocked":
            facts["session_started"] = True
            facts["status"] = "blocked"
    # Any post-launch record for a live flight proves the session is doing work → on the leg, not
    # merely just-airborne. (Applied after the specific handlers so a session_finished still wins
    # its BASE_TURN via report=True.)
    if act not in ("launch", "regenerate") and facts["launched"]:
        facts["session_started"] = True


def _stage(facts):
    launched = facts["launched"]
    return flights.flight_stage(
        facts["status"], report_present=facts["report"],
        session_started=facts["session_started"] and launched, launched=launched,
        cleared=facts["cleared"], closed=facts["closed"])


def _circuit_stage(facts):
    launched = facts["launched"]
    return flights.circuit_stage(
        facts["status"], report_present=facts["report"],
        session_started=facts["session_started"] and launched, launched=launched,
        cleared=facts["cleared"], closed=facts["closed"])


_OFF_PATH = (flights.PARKED, flights.AWAITING, flights.SESSION_FROZEN, flights.HOLDING)


def _flight_view(num, facts, tail):
    """One flight's frame object — the engine's model shape (design B.1: the pixels compute
    nothing), keyed exactly like the live field's so the replay binder maps them the same way."""
    stage = _stage(facts)
    contrail = "crisp" if stage in _TRAILING else "none"
    return {
        "num": num,
        "label": flights.flight_label(num, facts["attempt"]),
        "stage": stage,
        "circuit_stage": _circuit_stage(facts),
        "runway": num % 2,                     # no lane info in the journal — a stable parity runway
        "contrail": contrail,
        "spinning": False,                     # replay has no liveness axis — a treat, not a monitor
        "trouble": False,                      # no alarm dimming in the time-lapse; it stays clean
        "tail": tail,
    }


def _rec_num(rec):
    """The flight a record is about — its ``num``, else its ``id`` (``i23`` → 23), else an
    ``event.id`` envelope; ``None`` for a repo-wide record (a notify, an alert)."""
    n = rec.get("num")
    if isinstance(n, int) and not isinstance(n, bool):
        return n
    for src in (rec.get("id"), (rec.get("event") or {}).get("id")
                if isinstance(rec.get("event"), dict) else None):
        if isinstance(src, str) and src.startswith("i") and src[1:].isdigit():
            return int(src[1:])
    return None


def _valid_upto(journal, end):
    """Every valid record with a usable ts up to ``end`` (inclusive; ``None`` = unbounded), sorted
    ascending. A corrupt/absent ts can't sit on a timeline and is dropped, so the movie is always
    monotonic; non-dict lines are skipped (same rule as the tolerant journal reader). Records BEFORE
    the window's start are kept here on purpose — they carry the state the field already had when the
    window opened (see :func:`build_replay`)."""
    out = []
    for rec in journal:
        if not isinstance(rec, dict):
            continue
        ts = rec.get("ts")
        if not _finite(ts):
            continue
        if end is not None and ts > end:
            continue
        out.append(rec)
    out.sort(key=lambda r: r.get("ts"))
    return out


def build_replay(journal, *, slug="", name="", start=None, end=None,
                 hhmm=None, max_frames=_DEFAULT_MAX_FRAMES):
    """A journal (list of records) → the replay document the front-end scrubs.

    ``start``/``end`` (epoch seconds, inclusive, either ``None`` for unbounded) bound the window;
    ``hhmm`` formats a ts to the viewer's local ``HH:MM`` (injected so timezone never leaks into the
    lib); ``max_frames`` caps the payload, keeping the MOST RECENT frames and flagging truncation.

    The reconstruction walks EVERY valid record up to ``end`` so a windowed replay inherits the state
    the field already had when the window opened — a flight that launched an hour before a "last 6h"
    window is still in the air, not magically back at the stand. A FRAME is emitted only for records
    inside ``[start, end]``; earlier records are pre-roll that build state silently.

    Returns ``{slug, name, empty, window, frames}``. Each frame is one journal record applied to the
    cumulative reconstruction: ``{i, ts, hhmm, daypart, status, num, kind, radio, text, raw,
    flights: [...]}`` — the flights are every flight seen through this record, each at its discrete
    reconstructed stage, in the engine's model shape. ``num``/``raw`` make the frame a click target
    to its event."""
    hhmm = _local_hhmm if hhmm is None else hhmm
    tail = flights.airline_color(slug) if slug else None

    facts_by_num = {}
    order = []                                  # first-seen order, kept stable for the field list
    frames = []
    for rec in _valid_upto(journal, end):
        num = _rec_num(rec)
        if num is not None:
            if num not in facts_by_num:
                facts_by_num[num] = _new_facts()
                order.append(num)
            _apply(facts_by_num[num], rec)

        ts = rec.get("ts")
        if start is not None and ts < start:
            continue                            # pre-roll: accumulate state, emit no frame yet

        flight_views = [_flight_view(n, facts_by_num[n], tail) for n in order]
        status = "attention" if any(f["stage"] in _OFF_PATH for f in flight_views) else "ok"

        gloss = tower.comms_row(rec)
        frames.append({
            "ts": ts,
            "hhmm": hhmm(ts),
            "daypart": _daypart(ts),            # the living clock at the event's OWN time (§7), guarded
            "status": status,
            "num": num,
            "kind": gloss["kind"],
            "radio": gloss["radio"],
            "text": gloss["text"],
            "raw": json.dumps(rec, separators=(",", ":")),
            "flights": flight_views,
        })

    truncated = len(frames) > max_frames
    if truncated:
        frames = frames[-max_frames:]           # the recent past is what you replay
    for i, fr in enumerate(frames):
        fr["i"] = i

    return {
        "slug": slug,
        "name": name or slug,
        "empty": len(frames) == 0,
        "window": {
            "start": frames[0]["ts"] if frames else start,
            "end": frames[-1]["ts"] if frames else end,
            "frames": len(frames),
            "truncated": truncated,
        },
        "frames": frames,
    }

File: dashboard/static/replay.js
/* The night replay (Task 11 / design record §4) — a beloved TREAT, never load-bearing.
 *
 * window.CCReplay is a persistent overlay OUTSIDE #root (like the drawer/flag box) so the 2s poll
 * never touches it. It fetches server-derived FRAMES (/api/replay — every semantic pre-computed in
 * lib/replay: stage, contrail, lighting, the glossed sentence) and plays them back over the SAME
 * airfield engine the live field uses (window.AirfieldLive), scrubbable and steppable. A frame only
 * advances a plane when a real journal event changed its stage — the transit you watch IS the event.
 *
 * Design B.1: this file computes NO semantics. It maps the server's frame flights onto the engine's
 * model, drives scrub/step/play, and forwards a plane tap (or the caption) as `cc:drawer-open` so
 * every frame is clickable through to its event. It never derives a stage, a time, or a sentence. */
(function () {
  "use strict";

  function esc(s) {
    return String(s == null ? "" : s)
      .replace(/&/g, "&amp;").replace(/</g, "&lt;").replace(/>/g, "&gt;").replace(/"/g, "&quot;");
  }

  var node = null, canvas = null, engine = null;
  var frames = [], idx = 0, playing = false, timer = 0;
  var slug = "", fun = null;
  var STEP_MS = 900;                 // one frame per ~0.9s at play — a watchable time-lapse

  function el(id) { return document.getElementById(id); }

  function ensure() {
    if (node) return;
    node = document.createElement("div");
    node.id = "cc-replay";
    node.className = "cc-replay";
    node.innerHTML =
      '<div class="cc-replay-card">' +
        '<div class="cc-replay-head">' +
          '<span class="cc-replay-title">⏮ NIGHT REPLAY <b id="cc-replay-name"></b></span>' +
          '<span class="cc-replay-sub">a treat — not the answer to “what happened” (that’s the digest)</span>' +
          '<span class="cc-replay-trunc" id="cc-replay-trunc" hidden></span>' +
          '<select id="cc-replay-range" title="how far back to replay">' +
            '<option value="21600">last 6h</option>' +
            '<option value="43200">last 12h</option>' +
            '<option value="86400" selected>last 24h</option>' +
            '<option value="all">all history</option>' +
          '</select>' +
          '<button class="cc-replay-x" data-replay-close title="close (Esc)">✕</button>' +
        '</div>' +
        '<div class="cc-replay-stage"><canvas class="cc-replay-canvas"></canvas>' +
          '<div class="cc-replay-empty" id="cc-replay-empty" hidden>no journal events in this window</div>' +
        '</div>' +
        '<div class="cc-replay-caption" id="cc-replay-caption"></div>' +
        '<div class="cc-replay-controls">' +
          '<button class="cc-replay-btn" data-replay-step="-1" title="step back">◀</button>' +
          '<button class="cc-replay-btn play" id="cc-replay-play" title="play / pause">▶</button>' +
          '<button class="cc-replay-btn" data-replay-step="1" title="step forward">▶</button>' +
          '<input type="range" class="cc-replay-scrub" id="cc-replay-scrub" min="0" max="0" value="0">' +
          '<span class="cc-replay-pos" id="cc-replay-pos">0 / 0</span>' +
        '</div>' +
      '</div>';
    document.body.appendChild(node);
    canvas = node.querySelector(".cc-replay-canvas");

    node.addEventListener("click", function (e) {
      var t = e.target;
      if (!t || !t.closest) return;
      if (t === node || t.closest("[data-replay-close]")) { close(); return; }
      var step = t.closest("[data-replay-step]");
      if (step) { stepBy(Number(step.getAttribute("data-replay-step"))); return; }
      if (t.closest("#cc-replay-play")) { togglePlay(); return; }
      var cap = t.closest("[data-replay-num]");
      if (cap) { openEvent(Number(cap.getAttribute("data-replay-num"))); return; }
    });
    el("cc-replay-scrub").addEventListener("input", function () {
      pause(); idx = Number(this.value) || 0; renderFrame();
    });
    el("cc-replay-range").addEventListener("change", function () { load(this.value); });
    document.addEventListener("keydown", function (e) {
      if (!isOpen()) return;
      if (e.key === "Escape") { close(); return; }
      if (e.key === "ArrowLeft") { stepBy(-1); e.preventDefault(); }
      else if (e.key === "ArrowRight") { stepBy(1); e.preventDefault(); }
      else if (e.key === " ") { togglePlay(); e.preventDefault(); }
    });

    // A tapped plane in the replay opens that flight's drawer — the same event the live field fires,
    // so replay is clickable through to ground truth exactly like the airfield (tap-where-you-read).
    canvas.addEventListener("click", function (e) {
      if (!engine) return;
      var p = logical(e), num = engine.hitTest(p.x, p.y);
      if (num != null) openEvent(num);
    });
    canvas.addEventListener("mousemove", function (e) {
      if (!engine) return;
      var p = logical(e);
      canvas.style.cursor = engine.hitTest(p.x, p.y) != null ? "pointer" : "default";
    });
  }

  function logical(e) {
    var r = canvas.getBoundingClientRect();
    return { x: (e.clientX - r.left) * (400 / r.width), y: (e.clientY - r.top) * (270 / r.height) };
  }

  function isOpen() { return !!(node && node.classList.contains("open")); }

  function open(repoSlug, funMap) {
    ensure();
    slug = repoSlug || "";
    fun = funMap || null;
    el("cc-replay-name").textContent = slug;
    node.classList.add("open");
    if (!engine && window.AirfieldLive) engine = window.AirfieldLive.mount(canvas);
    load(el("cc-replay-range").value);
  }

  function load(range) {
    pause();
    el("cc-replay-caption").innerHTML = '<span class="loading">reading the journal…</span>';
    var url = "/api/replay?repo=" + encodeURIComponent(slug) + "&range=" + encodeURIComponent(range);
    fetch(url, { cache: "no-store" })
      .then(function (r) {
        // Read the body either way, but keep r.ok — a typed 500 (a bad journal line) must NOT render
        // as an empty replay; it shows the error honestly.
        return r.json().then(function (b) { return { ok: r.ok, body: b }; },
                             function () { return { ok: r.ok, body: null }; });
      })
      .then(function (res) {
        if (!res.ok || !res.body || res.body.error) {
          showError((res.body && res.body.error) || "replay unavailable");
          return;
        }
        setFrames(res.body);
      })
      .catch(function () { showError("couldn’t reach the command center"); });
  }

  function showError(msg) {
    frames = [];
    var empty = el("cc-replay-empty");
    empty.hidden = false;
    empty.textContent = msg;
    el("cc-replay-caption").innerHTML = '<span class="err">' + esc(msg) + '</span>';
    el("cc-replay-pos").textContent = "0 / 0";
    el("cc-replay-trunc").hidden = true;
    var scrub = el("cc-replay-scrub"); scrub.max = 0; scrub.value = 0;
    if (engine) engine.update({ resetKey: "replay-error", time: "day", status: "ok",
                                dim: false, banners: [], flights: [] });
  }

  function setFrames(rp) {
    frames = (rp && rp.frames) || [];
    var scrub = el("cc-replay-scrub"), empty = el("cc-replay-empty");
    empty.hidden = frames.length > 0;
    empty.textContent = "no journal events in this window";
    var trunc = el("cc-replay-trunc");
    if (rp.window && rp.window.truncated) {
      trunc.hidden = false;
      trunc.textContent = "· capped to the most recent " + frames.length + " frames";
    } else { trunc.hidden = true; }
    scrub.max = Math.max(0, frames.length - 1);
    idx = 0;
    if (!frames.length) {
      if (engine) engine.update({ resetKey: "replay-empty", time: "day", status: "ok", dim: false,
                                  banners: [], flights: [] });
      el("cc-replay-caption").innerHTML = "";
      el("cc-replay-pos").textContent = "0 / 0";
      return;
    }
    renderFrame();
  }

  // The engine model for a frame — the SAME shape field.js builds for the live field, mapping the
  // server's pre-derived flight fields onto the engine's keys (design B.1: no semantics here).
  function model(i) {
    var fr = frames[i];
    var airlines = !fun || fun.airlines !== false;
    var clock = !fun || fun.living_clock !== false;
    return {
      resetKey: "replay",
      time: clock ? (fr.daypart || "day") : "day",
      status: fr.status || "ok",
      dim: false,
      banners: [],       // replay scrubs recorded field states; it has never towed the name cloths

      flights: (fr.flights || []).map(function (f) {
        return { num: f.num, label: f.label, stage: f.stage, circuitStage: f.circuit_stage,
                 runway: f.runway || 0, contrail: f.contrail || "none",
                 spinning: !!f.spinning, trouble: !!f.trouble, tail: airlines ? f.tail : null };
      })
    };
  }

  function renderFrame() {
    if (!frames.length) return;
    idx = Math.max(0, Math.min(idx, frames.length - 1));
    if (engine) engine.update(model(idx));
    var fr = frames[idx];
    var radio = fr.radio ? '<span class="radio">' + esc(fr.radio) + '</span> ' : "";
    var chip = fr.num != null
      ? '<span class="fnum" data-replay-num="' + esc(fr.num) + '">SL-' + esc(fr.num) + '</span> ' : "";
    el("cc-replay-caption").innerHTML =
      '<span class="t">' + esc(fr.hhmm) + '</span>' + chip + radio +
      '<span class="msg">' + esc(fr.text) + '</span>' +
      (fr.num != null ? '<span class="hint">tap to open the flight →</span>' : "");
    el("cc-replay-scrub").value = idx;
    el("cc-replay-pos").textContent = (idx + 1) + " / " + frames.length;
  }

  function stepBy(d) { pause(); idx += d; renderFrame(); }

  function togglePlay() { if (playing) pause(); else play(); }

  function play() {
    if (!frames.length) return;
    if (idx >= frames.length - 1) idx = 0;         // replay from the top if parked at the end
    playing = true;
    el("cc-replay-play").textContent = "❚❚";
    window.clearInterval(timer);
    timer = window.setInterval(function () {
      if (idx >= frames.length - 1) { renderFrame(); pause(); return; }
      idx += 1; renderFrame();
    }, STEP_MS);
    renderFrame();
  }

  function pause() {
    playing = false;
    window.clearInterval(timer);
    var b = el("cc-replay-play");
    if (b) b.textContent = "▶";
  }

  function openEvent(num) {
    if (num == null) return;
    document.dispatchEvent(new CustomEvent("cc:drawer-open", { detail: { repo: slug, num: num } }));
  }

  function close() {
    pause();
    if (engine) { engine.destroy(); engine = null; }   // stop the rAF loop while the treat is put away
    if (node) node.classList.remove("open");
  }

  window.CCReplay = { open: open, isOpen: isOpen };
})();
